File: Yow.CoD.Finance.Domain/Model/AggregateRoot.cs
using System;
using System.Collections.Generic;
using Yow.CoD.Finance.Domain.Contracts;

namespace Yow.CoD.Finance.Domain.Model
{
    public abstract class AggregateRoot
    {
        private readonly List<Event> _uncommittedEvents = new List<Event>();
        private readonly Dictionary<Type, Action<object>> _handlers =new Dictionary<Type, Action<object>>();

        protected AggregateRoot(Guid id)
        {
            Id = id;
        }

        public Guid Id { get; }
        public int Version { get; private set; }

        public void ApplyEvent(Event payload)
        {
            _handlers[payload.GetType()](payload);
            Version++;
        }

        protected void AddEvent(Event uncommittedEvent)
        {
            _uncommittedEvents.Add(uncommittedEvent);
            ApplyEvent(uncommittedEvent);
        }

        protected void RegisterHandler<T>(Action<T> handler)
        {
            _handlers.Add(typeof(T), e=>handler((T)e));
        }

        public Event[] GetUncommittedEvents()
        {
            return _uncommittedEvents.ToArray();
        }

        public void ClearUncommittedEvents()
        {
            _uncommittedEvents.Clear();
        }
    }
}

File: typescript/packages/domain/src/model/aggregate-root.ts
import { Event } from "../contracts/event.js";

export abstract class AggregateRoot {
  readonly id: string;
  #uncommittedEvents: Event[] = [];
  #handlers = new Map<string, (event: Event) => void>();
  #version = 0;

  protected constructor(id: string) {
    this.id = id;
  }

  get version(): number {
    return this.#version;
  }

  applyEvent(event: Event): void {
    const handler = this.#handlers.get(event.eventType);
    if (!handler) {
      throw new Error(
        `No handler registered for event type '${event.eventType}'`,
      );
    }
    handler(event);
    this.#version++;
  }

  getUncommittedEvents(): Event[] {
    return [...this.#uncommittedEvents];
  }

  clearUncommittedEvents(): void {
    this.#uncommittedEvents = [];
  }

  protected addEvent(event: Event): void {
    this.applyEvent(event);
    this.#uncommittedEvents.push(event);
  }

  protected registerHandler<T extends Event>(
    eventType: string,
    handler: (event: T) => void,
  ): void {
    this.#handlers.set(eventType, handler as (event: Event) => void);
  }
}
